File: 2-数据处理与分析/数据分析/代码/第一问/LSTM.m
% 读取比赛数据
% 假设你有一个包含比赛详细数据的表，列名包括 'feature1', 'feature2', 'performance_label'
% 这里的 feature1、feature2 等代表你的比赛数据中的各种特征
% data = readtable('your_data.csv');

% 提取特征和标签
X = data{:, {'feature1', 'feature2'}};  % 选择你认为对表现有影响的特征
y = data{:, 'performance_label'};  % 'performance_label' 是你想要预测的表现

% 数据标准化
X_scaled = zscore(X);

% 划分训练集和测试集
rng(42);  % 设置随机种子以保证可复现性
[trainIdx, testIdx] = crossvalind('HoldOut', size(X_scaled, 1), 0.2);
X_train = X_scaled(trainIdx, :);
X_test = X_scaled(testIdx, :);
y_train = y(trainIdx);
y_test = y(testIdx);

% 构建循环神经网络模型
numFeatures = size(X_train, 2);
numHiddenUnits = 100;  % 可根据实际情况调整

% 构建神经网络层
layers = [ ...
    sequenceInputLayer(numFeatures)  % 输入层
    lstmLayer(numHiddenUnits, 'Activation', 'relu')  % LSTM层
    fullyConnectedLayer(1)  % 全连接层，用于输出
    regressionLayer  % 回归层，用于回归问题的输出
];

% 配置训练选项
options = trainingOptions('adam', ...
    'MaxEpochs', 10, ...
    'MiniBatchSize', 32, ...
    'Verbose', 1, ...
    'Plots', 'training-progress');

% 训练模型
net = trainNetwork(X_train', y_train', layers, options);

% 评估模型
predictions = predict(net, X_test');

% 计算均方误差
mse = mean((predictions - y_test').^2);
fprintf('Mean Squared Error on Test Data: %.4f\n', mse);

File: 2-数据处理与分析/数据分析/代码/第三问/LSTM/LSTM.py
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense

# 假设你有一个包含比赛详细数据的DataFrame，列名包括 'match_id', 'set', 'game', 'point', 'feature1', 'feature2', ...
# 这里的 feature1、feature2 等代表你的比赛数据中的各种特征

# 读取比赛数据
# df = pd.read_csv('your_data.csv')

# 构建训练数据和标签
X = df[['feature1', 'feature2', ...]]  # 选择你认为对表现有影响的特征
y = df['performance_label']  # 假设 'performance_label' 是你想要预测的表现

# 数据预处理
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# 为 LSTM 准备数据
timesteps = 1  # 可根据实际情况调整
features = X_train.shape[1]

X_train_lstm = X_train.reshape(X_train.shape[0], timesteps, features)
X_test_lstm = X_test.reshape(X_test.shape[0], timesteps, features)

# 构建 LSTM 模型
model = Sequential()
model.add(LSTM(units=100, activation='relu', input_shape=(timesteps, features)))
model.add(Dense(units=1))  # 输出层

# 编译模型
model.compile(optimizer='adam', loss='mean_squared_error')

# 查看模型的结构
model.summary()

# 训练模型
model.fit(X_train_lstm, y_train, epochs=10, batch_size=32)

# 评估模型
loss = model.evaluate(X_test_lstm, y_test)
print(f'Mean Squared Error on Test Data: {loss}')

# 进行预测
predictions = model.predict(X_test_lstm)

# 如果需要，你可以进一步对模型进行调优、添加更多特征，以提高模型性能
